include/robot_interfaces/robot_logger.hpp

```diff
@@ -0,0 +1,353 @@
+    }
+
+    /*
+    * append_field_data_to_robot_data() appends the data corresponding to
+    * every field at the same time index to the log file.
+    */
+
+    void append_field_data_to_file(std::vector<std::vector<double>> field_data)
+    {
+
+      infix_ostream_iterator<double> double_iterator(output_file_, ", ");
+
+      for (auto data : field_data)
+      {
+          std::copy(data.begin(),
+                    data.end() - 1,
+                    double_iterator);
+
+      }
+
+    }
+
+    static void *write(void *instance_pointer)
+    {
+        ((RobotLogger *)(instance_pointer))->write();
+        return nullptr;
+    }
+
+    /*
+    * write() is the master function which writes everything to the log file.
+    * It dumps all the data corresponding to block_size_ number of time indices
+    * at one go.
+    */
+
+    void write()
+    {
+
+        append_header_to_file();
+
+        while (!stop_was_called_ &&
+               !(logger_data_->desired_action->length() > 0))
+        {
+            real_time_tools::Timer::sleep_until_sec(0.1);
+        }
+
+        index_ = logger_data_->observation->newest_timeindex();
+
+        while (!stop_was_called_)
+        {
+            if (index_ + block_size_ <=
+                logger_data_->observation->newest_timeindex())
+            {
+#ifdef VERBOSE
+                auto t1 = std::chrono::high_resolution_clock::now();
+#endif
+
+                append_robot_data_to_file();
+
+                index_ += block_size_;
+
+#ifdef VERBOSE
+
+// to check whether the data being requested to be logged is in the buffer of
+// the timeseries and inspect effect of delays
+
+                std::cout << "Index trying to access, oldest index in the "
+                             "buffer: "
+                          << j << ","
+                          << logger_data_->observation->oldest_timeindex()
+                          << std::endl;
+
+                auto t2 = std::chrono::high_resolution_clock::now();
+                auto duration =
+                    std::chrono::duration_cast<std::chrono::microseconds>(t2 -
+                                                                          t1)
+                        .count();
+
+// to print the time taken for one block of data to be logged.
+
+                std::cout << "Time taken for one block of data to be logged: "
+                          << duration << std::endl;
+#endif
+            }
+        }
+
+    }
+
+    /*
+    * call start() to create the thread for the RobotLogger and start logging!
+    */
+
+    void start(std::string filename)
+    {
+
+        output_file_name_ = filename;
+        thread_->create_realtime_thread(&RobotLogger::write, this);
+
+    }
+
+    /*
+    * call stop() when you want to stop logging.
+    */
+
+    void stop()
+    {
+
+        output_file_.open(output_file_name_, std::ios_base::app);
+
+        append_robot_data_to_file();
+
+        output_file_.close();
+
+    }
```

Comment:
I just noticed, this does not actually stop the logging.  The `stop_was_called_ = true;` and `thread_->join();` of the destructor should be moved to here.

Reply:
Thanks, fixed.